atef/qt_helpers.py

```diff
@@ -49,78 +49,70 @@ def my_class:
-                NestedClass = QDataclassValue
-                dtype = object
-            elif List in normalized:
-                # Make sure we have list manipulation methods
-                NestedClass = QDataclassList
-                dtype = normalized[-1]
-            else:
-                NestedClass = QDataclassValue
-                dtype = normalized[-1]
-            setattr(
-                self,
-                field.name,
-                NestedClass.of_type(dtype)(
-                    data,
-                    field.name,
-                    parent=self,
-                ),
-            )
-
-
-normalize_map = {
-    'Optional': Optional,
-    'List': List,
-    'Number': float,
-    'int': int,
-    'str': str,
-    'bool': bool,
-    'Configuration': object,
-    'IdentifierAndComparison': object,
-    'Comparison': object,
-    'Severity': int,
-    'reduce.ReduceMethod': str,
-    'PrimitiveType': object,
-    'Sequence': List,
-    'Value': object,
-    'Dict': Dict,
-    'str, Any': object,
-    'str, List': object,
-    'GroupResultMode': str,
-    'tools.Tool': object,
-}
-
-
-def normalize_annotation(annotation: str) -> Tuple[type]:
-    """
-    Change a string annotation into a tuple of the enclosing classes.
+        fields = get_type_hints(type(data))
+        for name, type_hint in fields.items():
+            self.set_field_from_data(name, type_hint, data)
 
-    For example: "Optional[List[SomeClass]]" becomes
-    (Optional, List, object)
-
-    Somewhat incomplete- only correct to the level needed for this
-    application.
-
-    Only supports the cases where we have exactly one element in each
-    square bracket nesting level.
+    def set_field_from_data(
+        self,
+        name: str,
+        type_hint: Any,
+        data: Any
+    ):
+        """
+        Set a field for this bridge based on the data and its type
 
-    There is definitely a better way to handle this, but I can't
-    figure it out quickly and want to press forward to v0.
-    """
-    elems = []
-    for text in annotation.strip(']').split('['):
-        elems.append(normalize_map[text])
-    return tuple(elems)
+        Parameters
+        ----------
+        name : str
+            name of the field
+        type_hint : Any
+            The type hint annotation, returned from typing.get_type_hints
+        data : any
+            The dataclass for this bridge
+        """
+        # Need to figure out which category this is:
+        # 1. Primitive value -> make a QDataclassValue
+        # 2. Another dataclass -> make a QDataclassValue (object)
+        # 3. A list of values -> make a QDataclassList
+        # 4. A list of dataclasses -> QDataclassList (object)
+        origin = get_origin(type_hint)
+        args = get_args(type_hint)
+
+        if not origin:
+            # a raw type, no Union, Optional, etc
+            NestedClass = QDataclassValue
+            dtype = type_hint
+        elif origin == dict:
```

Comment:
I would always choose `type_a is type_b` in these scenarios (instead of `==`). Not strictly necessary to change, though, as the end result is the same for these builtins.